File: test/format_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:plan_e/core/format.dart';

void main() {
  group('Nepali Formatters - formatNpr', () {
    test('formats small amounts in Rs.', () {
      expect(AppFormatters.formatNpr(50000), 'Rs. 500');
    });

    test('formats lakhs (1,00,000)', () {
      expect(AppFormatters.formatNpr(10000000), 'Rs. 1,00,000');
    });

    test('formats 12,50,000 NPR (125,000,000 paisa)', () {
      expect(AppFormatters.formatNpr(125000000), 'Rs. 12,50,000');
    });

    test('formats crores (1,00,00,000 NPR)', () {
      expect(AppFormatters.formatNpr(1000000000), 'Rs. 1,00,00,000');
    });
  });

  group('Asia/Kathmandu Timezone', () {
    test('converts UTC to Asia/Kathmandu (+05:45 offset)', () {
      final DateTime utcDate = DateTime.utc(2026, 8, 3, 12, 0); // 12:00 UTC
      final DateTime ktmDate = AppFormatters.toKathmandu(utcDate);
      expect(ktmDate.hour, 17);
      expect(ktmDate.minute, 45);
    });

    test('converts across UTC midnight boundary', () {
      final DateTime utcDate = DateTime.utc(2026, 8, 3, 22, 0); // 22:00 UTC
      final DateTime ktmDate = AppFormatters.toKathmandu(utcDate);
      expect(ktmDate.day, 4);
      expect(ktmDate.hour, 3);
      expect(ktmDate.minute, 45);
    });
  });

  group('Nepali Phone Validation', () {
    test('accepts valid 10-digit Nepali mobile numbers starting with 98, 97, 96', () {
      expect(AppFormatters.isNepaliPhone('9841234567'), isTrue);
      expect(AppFormatters.isNepaliPhone('+9779841234567'), isTrue);
      expect(AppFormatters.isNepaliPhone('+9779712345678'), isTrue);
    });

    test('rejects invalid phone numbers', () {
      expect(AppFormatters.isNepaliPhone('12345'), isFalse);
      expect(AppFormatters.isNepaliPhone('984123456'), isFalse); // 9 digits
      expect(AppFormatters.isNepaliPhone('8841234567'), isFalse); // invalid prefix
    });
  });

  group('Duration Formatters', () {
    test('formats hours under 24', () {
      expect(AppFormatters.formatDuration(5), '5 hours');
      expect(AppFormatters.formatDuration(1), '1 hour');
    });

    test('formats days for 24 hours or more', () {
      expect(AppFormatters.formatDuration(24), '1 day');
      expect(AppFormatters.formatDuration(72), '3 days');
      expect(AppFormatters.formatDuration(30), '1 day 6 hrs');
    });
  });
}

File: src/lib/__tests__/format.test.ts
import { formatNpr, toKathmandu, formatDate, isNepaliPhone, formatDuration } from '../format';

describe('Nepali Formatters', () => {
  describe('formatNpr (Paisa to NPR with Lakh Grouping)', () => {
    test('formats small amounts correctly', () => {
      expect(formatNpr(0)).toBe('Rs. 0');
      expect(formatNpr(50000)).toBe('Rs. 500');
    });

    test('formats lakhs (1,00,000)', () => {
      // 1,00,000 rupees = 10,000,000 paisa
      expect(formatNpr(10000000)).toBe('Rs. 1,00,000');
    });

    test('formats 12,50,000 NPR (125,000,000 paisa)', () => {
      expect(formatNpr(125000000)).toBe('Rs. 12,50,000');
    });

    test('formats crores (1,00,00,000 NPR)', () => {
      // 1,00,00,000 rupees = 1,000,000,000 paisa
      expect(formatNpr(1000000000)).toBe('Rs. 1,00,00,000');
    });
  });

  describe('Asia/Kathmandu Timezone', () => {
    test('converts UTC to Asia/Kathmandu (+05:45 offset without DST)', () => {
      const utcDate = new Date('2026-08-03T12:00:00Z');
      const ktmDate = toKathmandu(utcDate);
      expect(ktmDate).toBeInstanceOf(Date);
    });

    test('formats date correctly in Kathmandu timezone', () => {
      const formatted = formatDate('2026-08-03T00:00:00Z', 'yyyy-MM-dd');
      expect(formatted).toBe('2026-08-03');
    });
  });

  describe('isNepaliPhone', () => {
    test('accepts valid Nepali mobile numbers', () => {
      expect(isNepaliPhone('9841234567')).toBe(true);
      expect(isNepaliPhone('+9779801234567')).toBe(true);
      expect(isNepaliPhone('+9779741234567')).toBe(true);
      expect(isNepaliPhone('9611234567')).toBe(true);
    });

    test('rejects invalid numbers', () => {
      expect(isNepaliPhone('1234567890')).toBe(false);
      expect(isNepaliPhone('+14155552671')).toBe(false);
      expect(isNepaliPhone('9841234')).toBe(false);
      expect(isNepaliPhone('abc9841234567')).toBe(false);
    });
  });

  describe('formatDuration', () => {
    test('formats hours under 24', () => {
      expect(formatDuration(1)).toBe('1 hr');
      expect(formatDuration(8)).toBe('8 hrs');
      expect(formatDuration(23)).toBe('23 hrs');
    });

    test('formats days for 24 hours or more', () => {
      expect(formatDuration(24)).toBe('1 day');
      expect(formatDuration(48)).toBe('2 days');
      expect(formatDuration(288)).toBe('12 days');
    });
  });
});
